plugin/session_buffer.py

```diff
@@ -408,5 +428,113 @@ def _present_diagnostics_async(
+        for idx in range(0, len(self.semantic_tokens.data), 5):
+            delta_line = self.semantic_tokens.data[idx]
+            delta_start_utf16 = self.semantic_tokens.data[idx + 1]
+            length_utf16 = self.semantic_tokens.data[idx + 2]
+            token_type_encoded = self.semantic_tokens.data[idx + 3]
+            token_modifiers_encoded = self.semantic_tokens.data[idx + 4]
+            line = prev_line + delta_line
+            col_utf16 = prev_col_utf16 + delta_start_utf16 if delta_line == 0 else delta_start_utf16
+            a = view.text_point_utf16(line, col_utf16, clamp_column=False)
+            b = view.text_point_utf16(line, col_utf16 + length_utf16, clamp_column=False)
+            r = sublime.Region(a, b)
+            prev_line = line
+            prev_col_utf16 = col_utf16
+            token_type, token_modifiers, scope = self.session.decode_semantic_token(
+                token_type_encoded, token_modifiers_encoded)
+            if scope is None:
+                # We can still use the meta scope and draw highlighting regions for custom token types if there is a
+                # color scheme rule for this particular token type.
+                # This logic should not be cached (in the _decode_semantic_token method) because otherwise new user
+                # customizations in the color scheme for the scopes of custom token types would require a restart of
+                # Sublime Text to take effect.
+                token_general_style = view.style_for_scope("meta.semantic-token")
+                token_type_style = view.style_for_scope("meta.semantic-token.{}".format(token_type.lower()))
+                if token_general_style["source_line"] != token_type_style["source_line"] or \
+                        token_general_style["source_column"] != token_type_style["source_column"]:
+                    scope = "meta.semantic-token.{}.lsp".format(token_type.lower())
+            self.semantic_tokens.tokens.append(SemanticToken(r, token_type, token_modifiers))
+            if scope:
+                scope_regions.setdefault(scope, []).append(r)
+        # don't update regions if there were additional changes to the buffer in the meantime
+        if self.semantic_tokens.view_change_count != view.change_count():
+            return
+        removed_scopes = []
+        for scope in self.semantic_tokens.active_scopes.keys():
+            if scope not in scope_regions.keys():
+                removed_scopes.append(scope)
+                for sv in self.session_views:
+                    sv.view.erase_regions("lsp_{}".format(scope))
+        for scope in removed_scopes:
+            del self.semantic_tokens.active_scopes[scope]
+        for scope, regions in scope_regions.items():
+            regions_hash = hash(tuple(hash((r.a, r.b)) for r in regions))
+            if scope not in self.semantic_tokens.active_scopes or \
+                    self.semantic_tokens.active_scopes[scope] != regions_hash:
```

Comment:
It looks like this optimization might be doing a wrong thing when a single character is replaced (either in place or just quickly enough).

https://user-images.githubusercontent.com/153197/144705583-e22813b2-90f5-4d2b-b7ac-8e586199cf9d.mov

Reply:
Yeah, looks problematic, thanks for finding this bug. I've removed this check/optimization to redraw all regions after each new response in 80feebd. The alternative would be to include the region's content into the hash value as well, but I guess all the `View.substr()` calls in that case might be more expensive than just to redraw all regions, even if unchanged.